app/uk/gov/hmrc/fileupload/controllers/envelope.scala

```diff
@@ -47,21 +49,51 @@ object EnvelopeReport {
       expiryDate = envelope.expiryDate,
       status = Some(envelope.status.name),
       metadata = envelope.metadata,
+      constraints = envelope.constraints,
       destination = envelope.destination,
       application = envelope.application,
       files = fileReports
     )
   }
-
 }
 
 case class CreateEnvelopeRequest(callbackUrl: Option[String] = None,
                                  expiryDate: Option[DateTime] = None,
-                                 metadata: Option[JsObject] = None)
+                                 metadata: Option[JsObject] = None,
+                                 constraints: Option[EnvelopeConstraintsO] = None)
+
+case class EnvelopeConstraintsO(maxNumFiles: Option[Int] = None,
+                                maxSize: Option[String] = None,
+                                maxSizePerItem: Option[String] = None)
+
+case class EnvelopeConstraints(maxNumFiles: Int,
+                               maxSize: String,
+                               maxSizePerItem: String)
+
 
 object CreateEnvelopeRequest {
   implicit val dateReads = Reads.jodaDateReads("yyyy-MM-dd'T'HH:mm:ss'Z'")
+  implicit val constraintsFormats = Json.format[EnvelopeConstraintsO]
   implicit val formats = Json.format[CreateEnvelopeRequest]
+
+  def constraintsWithDefaultsIfNotProvided(constraints: Option[EnvelopeConstraintsO], defaultConstraints: EnvelopeConstraints): EnvelopeConstraints = {
+    val defaultMaxCapacity = defaultConstraints.maxNumFiles
+    val defaultMaxSize = defaultConstraints.maxSize
+    val defaultMaxSizePerItem = defaultConstraints.maxSizePerItem
+
+    constraints match {
+      case Some(constraints) => {
```

Comment:
using same `constraints` naming shadows the original value that you are pattern matching on.